### ISSUE TITLE
Results not consistent with previous build?

### ISSUE THREAD

**Issue by cciccole** [CannyLab/tsne-cuda#42]
I can't seem to produce good embeddings with v2. Check out the difference on the same dataset with an old build versus the new one:

## V2 build:
![newbuild](https://user-images.githubusercontent.com/3837064/59534940-7e85f600-8ea4-11e9-97ea-203b27410cb0.png)

### V2 build logs:
<details>
[Step 0] Avg. Gradient Norm: 0.000577044

[Step 50] Avg. Gradient Norm: 0.0707002

[Step 100] Avg. Gradient Norm: 0.0651156

[Step 150] Avg. Gradient Norm: 0.0585839

[Step 200] Avg. Gradient Norm: 0.0517932

[Step 250] Avg. Gradient Norm: 0.0723898

[Step 300] Avg. Gradient Norm: 0.0590655

[Step 350] Avg. Gradient Norm: 0.0462633

[Step 400] Avg. Gradient Norm: 0.0449879

[Step 450] Avg. Gradient Norm: 0.0418904

[Step 500] Avg. Gradient Norm: 0.0390911

[Step 550] Avg. Gradient Norm: 0.0356273

[Step 600] Avg. Gradient Norm: 0.0310828

[Step 650] Avg. Gradient Norm: 0.0245189

[Step 700] Avg. Gradient Norm: 0.0188884

[Step 750] Avg. Gradient Norm: 0.017637

[Step 800] Avg. Gradient Norm: 0.0151654

[Step 850] Avg. Gradient Norm: 0.0118614

[Step 900] Avg. Gradient Norm: 0.0126098

[Step 950] Avg. Gradient Norm: 0.0119634 
</details>

## Version efa209834879bba88814e74d7062539f4de07cc2
![oldbuild](https://user-images.githubusercontent.com/3837064/59535043-bd1bb080-8ea4-11e9-8d71-cb4d3f78686b.png)

### Old version logs:
<details>
 [Step 0] Avg. Gradient Norm: 5.4169e-06

[Step 50] Avg. Gradient Norm: 0.0909963

[Step 100] Avg. Gradient Norm: 0.0573804

[Step 150] Avg. Gradient Norm: 0.0432024

[Step 200] Avg. Gradient Norm: 0.0329338

[Step 250] Avg. Gradient Norm: 0.0818289

[Step 300] Avg. Gradient Norm: 0.0811468

[Step 350] Avg. Gradient Norm: 0.0756753

[Step 400] Avg. Gradient Norm: 0.0717059

[Step 450] Avg. Gradient Norm: 0.0687041

[Step 500] Avg. Gradient Norm: 0.0663398

[Step 550] Avg. Gradient Norm: 0.0642467

[Step 600] Avg. Gradient Norm: 0.0623364

[Step 650] Avg. Gradient Norm: 0.0605533

[Step 700] Avg. Gradient Norm: 0.0588856

[Step 750] Avg. Gradient Norm: 0.0572197

[Step 800] Avg. Gradient Norm: 0.0553893

[Step 850] Avg. Gradient Norm: 0.05345

[Step 900] Avg. Gradient Norm: 0.051608

[Step 950] Avg. Gradient Norm: 0.0497179

</details>

The reason the old build is just a git tag is because I built it many months ago when master was at that point. V2 is compiled from source from the V2 tag. Both were compiled in an ubuntu 16.04 / cuda9.0 environment but not the exact same environment for both.

The settings for the run are the same between builds. The differences from default settings are `early exag = 4`, `learning rate = 20000`, `n neighbors = 100`. The dataset is 500k rows and 14 columns. I can provide it if you want to try it out. That might be the simplest way to judge if it's some build issue on my end.

Do you have any standard datasets that you run through the algorithm as a quality check? Have you noticed any differences between your old build and the new one? If any, I'm sure they aren't this bad. Is it possible that the characteristics of this data are bringing out an edge case?

**Comment by rmrao** [comment]
That is... odd. We've run this on multiple standard datasets (MNIST, CIFAR, etc.) and used it in some of our other work ourselves for months. It's believable to me that the learning rates, etc. required might have changed, but we've regularly gotten good results with it. If you could provide the dataset I can try and debug. Not sure how big the file is but you can email it to me (roshan_rao@berkeley.edu) or put it on AWS/some other file sharing service if its large.

**Comment by cciccole** [comment]
Here is the dataset (65MB) https://drive.google.com/file/d/16NBS3Pf-14_HfYoH-WNx1ytTiCJZ63vV/view?usp=sharing

If you could give it a spin and see if it's just my build that would be useful. Else maybe it will help uncover the source of the problem causing this, if it is indeed a problem.

I tried playing with the nearest neighbors, perplexity, learning rate, and early exaggeration factor but couldn't find anything that worked. Choices of these parameters that I know to be bad did produce terrible embeddings with no discernible structure at all. However, I could only achieve the most basic directional improvement in the result like what you see above. I should probably try theta as well. Maybe you also have some other ideas. Did you find that your testing of the standard datasets required any different choice of parameters? Or did established conventions hold up?

**Comment by rmrao** [comment]
Ah - there's a bug in the apply forces kernel. I'm not sure how it worked for us, maybe when I rebuilt things and merged something got overwritten. Will fix that and hopefully it will solve your problem. Might take a day or two to test the fix and recompile.

**Comment by rmrao** [comment]
@DavidMChan We should filter out all uses of `num_nodes` now since it's not used and causes bugs.

**Comment by cciccole** [comment]
Wow that was fast. Sounds good I'll look out for more info from you.

On the subject of compilation, it seems to me that multicore compilation still is error-prone. I tried it with 24 cores on a 32 core machine. Do you have a safe number for parallel compilation that seems to work for you consistently?

**Comment by rmrao** [comment]
Well... discovered a handful of bugs while testing this. First is that the `early_exaggeration` argument does nothing in the python extension - we weren't passing it through. Since a lot of our tests on large datasets were run with the C extension and early exaggeration doesn't affect smaller datasets as much, we probably didn't notice.

I've gotten results that look much better after fixing this, although I had to mess with the parameters a little more. I'm going to continue testing and double checking that there isn't another bug somewhere.

**Comment by rmrao** [comment]
Ok, it looks like the fit-tsne code defaults to an early exaggeration of 12 (see [here](https://github.com/KlugerLab/FIt-SNE/blob/5774b0781f10bd9d526d6a830091b3624a8b1232/fast_tsne.py#L21)). 

Since this is the default exaggeration and you doubled the bh_tsne default exaggeration, I ran it with a doubled fit_tsne exaggeration. So the parameters I used were `early_exaggeration=12, learning_rate=20000, num_neighbors=100`. After fixing the early exaggeration bug I get this:

<img width="1346" alt="image" src="https://user-images.githubusercontent.com/6496605/59557085-502c1780-8f85-11e9-9c21-be0973915d92.png">

which seems consistent with your results on bh-tsne.

**Comment by rmrao** [comment]
Also re: parallel compilation... I have no idea what the issue is. I'm able to run it with 32 threads on a 12 core machine without any issues. Except that when I use `conda-build` I have compilation issues.

It's possible that conda uses a different compiler by default (we were using `gcc-4.9`, but I tried disabling that and still had issues). I'm able to consistently parallel compile without issues on Ubuntu 18.04 with `gcc-7.4`.

**Comment by rmrao** [comment]
Ok fix was merged to master in #45. I've opened a new issue for the compilation problem (#46).

Let me know if you still have issues.

**Comment by cciccole** [comment]
Okay so I tried current master and I do see the improvement. However, if I use a [slightly larger dataset](http://www.omiq.ai/data/opt-sne/mass41_for_optsne.csv.zip) that is mostly composed of the same data as before, the result begins to deteriorate again. The issue is a high amount of noise in the embedding, for example:

![image](https://user-images.githubusercontent.com/3837064/59670294-8f6a8c00-9170-11e9-9336-03232a3da3e1.png)

This pattern of having a clear structure present but then a bunch of noise scattered everywhere around was the same as before, except it was much worse before.

The same dataset with BH version:
![image](https://user-images.githubusercontent.com/3837064/59670784-5848aa80-9171-11e9-83a5-9663ec2ba493.png)

I tried different combinations of settings. The results above are representative.

I link the dataset above. Maybe you can give that one a try now?

**Comment by rmrao** [comment]
Well I played around with this a little bit today... and I can't seem to find what the issue is easily. 

If I had to make a *guess* (and it's just a guess at this point), there might be an issue with the repulsive forces kernel, or with normalization computation. At least the normalization is something like 2e9 to start with and doesn't seem to go down for a while (600-700 iterations). That being said I'm not at all sure that's the issue.

We also changed the attractive forces kernel to an atomicAdd instead of the cusparse implementation... that may have introduced something?

Double checking results on other datasets, this phenomenon does not affect all datasets and really only seems to be a big issue when you start increasing dataset size (which obviously is problematic since this is a major use case of this repository). I'll have to test the two version side by side and see if I can find the cause.

**Comment by rmrao** [comment]
I've officially determined that this is a bug in our implementation of the FIt-SNE repulsive forces.

Undoing changes to attractive forces doesn't have any effect on the resulting embedding. This *mainly* seems to stem from normalization the forces, and no difference is evident for datasets of size <= 100k points, with serious issues cropping up in datasets of size >= 500k points.

Will need to investigate further, and the repulsive force code is quite complicated so it may take a while to fully debug. If it's really a pressing issue for someone, please revert to the `bh_tsne` implementation (either the branch on or the tag posted above in this issue). Hopefully we can get this fixed soon.

**Comment by rmrao** [comment]
So at commit a0eb8d0dea769b36c9a65c0e453150dea4bbdc71, it looks like things work as expected.

This is about 1/2-3/4 of the way through our FIt-SNE conversion. At this point in the history, a handful of easier things are still on CPU (we did the hard conversions first). Additionally, there are some optimizations we experimented with later on that may have caused the issue. However there's now a clear starting point where the bug did not exist, so we should be able to roll forward until we find the problem.

**Comment by cciccole** [comment]
Thanks for the update. Maybe you can setup an automated build and test process with git-bisect or similar to make this easier.

**Comment by rmrao** [comment]
Well this is probably one of the most painful bugs we've encountered, but I believe we have resolved it in #51. Here is the dataset you posted with parameters: `early_exaggeration=4, learning_rate=20000, num_neighbors=100`:

![image](https://user-images.githubusercontent.com/6496605/60848477-a0208780-a19b-11e9-9a11-d1f547778270.png)

I'll leave this issue open for a bit. Let us know if there are any further problems with your datasets.

**Comment by cciccole** [comment]
Nice, glad you all sorted it out, or at least it seems so to me. 

This issue is now closed and that's fine. If I notice anything strange I'll say so.

**Comment by cciccole** [comment]
@rmrao when will this current build be available on conda?

**Comment by rmrao** [comment]
Probably soon. We’re working on creating a docket environment to build all of the conda versions since the current build process is quite painful. Hopefully we can add a continuous integration build so updates will make it to conda more quickly in the future.

**Comment by rmrao** [comment]
Up on conda now. Compatible with cuda versions 9.0, 9.1, 9.2, 10.0, and 10.1.
